diff -u /home/southerd/Desktop/sqlite3//sqlite3_driver.php /home/southerd/Devel/southerd/webdesk/dev/webdesk/app/www/system/database/drivers/sqlite3//sqlite3_driver.php
--- /home/southerd/Desktop/sqlite3//sqlite3_driver.php	2010-08-22 19:03:07.000000000 -0600
+++ /home/southerd/Devel/southerd/webdesk/dev/webdesk/app/www/system/database/drivers/sqlite3//sqlite3_driver.php	2010-11-17 17:14:03.536666734 -0700
@@ -19,7 +19,7 @@
 
 
 /**
- * SQLite Database Adapter Class
+ * SQLite3 Database Adapter Class
  *
  * Note: _DB is an extender class that the app controller
  * creates dynamically based on whether the active record
@@ -33,10 +33,10 @@
  */
 class CI_DB_sqlite3_driver extends CI_DB {
 
-	var $dbdriver = 'sqlite';
+	var $dbdriver = 'sqlite3';
 	
-	// The character used to escape with - not needed for SQLite
-	var $_escape_char = '';
+	// The character used for escaping
+	var	$_escape_char = '`';
 
 	// clause and character used for LIKE escape sequences
  	var $_like_escape_str = " ESCAPE '%s' ";
@@ -50,6 +50,8 @@
 	var $_count_string = "SELECT COUNT(*) AS ";
 	var $_random_keyword = ' Random()'; // database specific random keyword
 
+	var $_results = array(); //a private array to cache results.
+
 	/**
 	 * Non-persistent database connection
 	 *
@@ -86,7 +88,8 @@
 	 */	
 	function db_pconnect()
 	{
-		// Not supported
+		// The database is persistent. There is no difference between a normal
+		// and a persistent connection.
 		return $this->db_connect();
 	}
 	
@@ -103,7 +106,12 @@
 	 */
 	function reconnect()
 	{
-		// not implemented in SQLite
+		// Not needed in SQLite
+		// Specifically, timeout should be infinite. Still, would be nice to 
+		// close the connection.
+		$this->close();
+		$this->db_connect(); //$this->database hasn't been changed.
+		                     // Or maybe it has, which is why we're reconnecting.
 	}
 
 	// --------------------------------------------------------------------
@@ -160,9 +168,8 @@
 	 */	
 	function _execute($sql)
 	{
-		static $results = array(); // FIXME: some kind of problems happen with the error messages/error code when the results are picked by GC
 		$sql = $this->_prep_query($sql);
-		return $results[] = @$this->conn_id->query($sql);
+		return $this->conn_id->query($sql);
 	}
 	
 	// --------------------------------------------------------------------
@@ -641,12 +648,15 @@
 	 */
 	function _close($conn_id)
 	{
-		@$conn_id->close();
+		$conn_id->close();
+		if($conn_id == $this->conn_id){
+			$this->conn_id = FALSE;
+		}
 	}
 
 
 }
 
 
-/* End of file sqlite_driver.php */
-/* Location: ./system/database/drivers/sqlite/sqlite_driver.php */
+/* End of file sqlite3_driver.php */
+/* Location: ./system/database/drivers/sqlite3/sqlite3_driver.php */
diff -u /home/southerd/Desktop/sqlite3//sqlite3_forge.php /home/southerd/Devel/southerd/webdesk/dev/webdesk/app/www/system/database/drivers/sqlite3//sqlite3_forge.php
--- /home/southerd/Desktop/sqlite3//sqlite3_forge.php	2010-08-22 14:17:17.000000000 -0600
+++ /home/southerd/Devel/southerd/webdesk/dev/webdesk/app/www/system/database/drivers/sqlite3//sqlite3_forge.php	2010-11-17 16:11:24.686666733 -0700
@@ -260,4 +260,4 @@
 }
 
 /* End of file sqlite3_forge.php */
-/* Location: ./system/database/drivers/sqlite/sqlite3_forge.php */
+/* Location: ./system/database/drivers/sqlite3/sqlite3_forge.php */
Only in /home/southerd/Desktop/sqlite3/: sqlite3.patch
diff -u /home/southerd/Desktop/sqlite3//sqlite3_result.php /home/southerd/Devel/southerd/webdesk/dev/webdesk/app/www/system/database/drivers/sqlite3//sqlite3_result.php
--- /home/southerd/Desktop/sqlite3//sqlite3_result.php	2010-08-22 13:16:00.000000000 -0600
+++ /home/southerd/Devel/southerd/webdesk/dev/webdesk/app/www/system/database/drivers/sqlite3//sqlite3_result.php	2010-11-17 16:31:25.430000066 -0700
@@ -25,18 +25,10 @@
  * @link		http://codeigniter.com/user_guide/database/
  */
 class CI_DB_sqlite3_result extends CI_DB_result {
-	var $data = NULL;
-	var $pointer = 0;
+	var $_data = NULL;
+	var $_pointer = 0;
+	var $_num_rows = -1;
 	
-	function _set_data() {
-		if ($this->data === NULL) {
-			$this->data = array();
-			while ($row = $this->result_id->fetchArray(SQLITE3_ASSOC)) {
-				$this->data[] = $row;
-			}
-		}
-	}
-
 	/**
 	 * Number of rows in the result set
 	 *
@@ -45,8 +37,13 @@
 	 */
 	function num_rows()
 	{
-		$this->_set_data();
-		return count((array)@$this->data);
+		if($this->_data === NULL)
+		{
+			$this->_set_data();
+			$this->_num_rows = count($this->_data);
+		}
+				
+		return $this->_num_rows;
 	}
 	
 	// --------------------------------------------------------------------
@@ -59,7 +56,7 @@
 	 */
 	function num_fields()
 	{
-		return @$this->result_id->numColumns();
+		return $this->result_id->numColumns();
 	}
 
 	// --------------------------------------------------------------------
@@ -120,7 +117,27 @@
 	 */		
 	function free_result()
 	{
-		// Not implemented in SQLite
+		$this->_data = NULL;
+		$this->_pointer = 0;
+		$this->_num_rows = -1;
+	}
+
+	/**
+	 * Load the result set into an internal array.
+	 *
+	 * @return	null
+	 */		
+
+	function _set_data()
+	{
+		if ($this->_data === NULL)
+		{
+			$this->_data = array();
+			while ($row = $this->result_id->fetchArray(SQLITE3_ASSOC))
+			{
+				$this->_data[] = $row;
+			}
+		}
 	}
 
 	// --------------------------------------------------------------------
@@ -153,7 +170,7 @@
 	function _fetch_assoc()
 	{
 		$this->_set_data();
-		return @$this->data[$this->pointer++];
+		return @$this->_data[$this->_pointer++];
 	}
 	
 	// --------------------------------------------------------------------
@@ -177,9 +194,8 @@
 			return NULL;
 		}
 	}
-
 }
 
 
 /* End of file sqlite3_result.php */
-/* Location: ./system/database/drivers/sqlite/sqlite3_result.php */
+/* Location: ./system/database/drivers/sqlite3/sqlite3_result.php */
diff -u /home/southerd/Desktop/sqlite3//sqlite3_utility.php /home/southerd/Devel/southerd/webdesk/dev/webdesk/app/www/system/database/drivers/sqlite3//sqlite3_utility.php
--- /home/southerd/Desktop/sqlite3//sqlite3_utility.php	2010-08-22 12:04:39.000000000 -0600
+++ /home/southerd/Devel/southerd/webdesk/dev/webdesk/app/www/system/database/drivers/sqlite3//sqlite3_utility.php	2010-11-17 17:08:53.966666734 -0700
@@ -27,21 +27,15 @@
 	/**
 	 * List databases
 	 *
-	 * I don't believe you can do a database listing with SQLite
-	 * since each database is its own file.  I suppose we could
-	 * try reading a directory looking for SQLite files, but
-	 * that doesn't seem like a terribly good idea
+	 * There is only one database ever loaded, so we just return the currently
+	 * loaded database.
 	 *
 	 * @access	private
 	 * @return	bool
 	 */
 	function _list_databases()
 	{
-		if ($this->db_debug)
-		{
-			return $this->display_error('db_unsuported_feature');
-		}
-		return array();
+		return "SELECT db_name()";
 	}
 
 	// --------------------------------------------------------------------
@@ -49,7 +43,8 @@
 	/**
 	 * Optimize table query
 	 *
-	 * Is optimization even supported in SQLite?
+	 * SQLite table optimization is the VACUUM command. See
+	 * http://www.sqlite.org/lang_vacuum.html
 	 *
 	 * @access	private
 	 * @param	string	the table name
@@ -57,7 +52,7 @@
 	 */
 	function _optimize_table($table)
 	{
-		return FALSE;
+		return "VACUUM";
 	}
 
 	// --------------------------------------------------------------------
@@ -65,7 +60,7 @@
 	/**
 	 * Repair table query
 	 *
-	 * Are table repairs even supported in SQLite?
+	 * SQLite "repairs" automatically on opening the database.
 	 *
 	 * @access	private
 	 * @param	string	the table name
@@ -87,8 +82,13 @@
 	 */
 	function _backup($params = array())
 	{
-		// Currently unsupported
-		return $this->db->display_error('db_unsuported_feature');
+		$dumpresult = false;
+		//It would be really, really nice if it worked this way.		
+		//Sadly, .dump is a sqlite shell builtin, and I don't feel like copying
+		//it, or shoehorning the mysql backup query.
+/*		$sql = ".dump";
+		$dumpresult = $this->db->query($sql);*/
+		return $dumpresult;
 	}
 
 	/**
@@ -138,4 +138,4 @@
 }
 
 /* End of file sqlite3_utility.php */
-/* Location: ./system/database/drivers/sqlite/sqlite3_utility.php */
+/* Location: ./system/database/drivers/sqlite3/sqlite3_utility.php */
